File: android/app/src/test/kotlin/is/borgarland/poc/FactsFileTest.kt
package `is`.borgarland.poc

import `is`.borgarland.poc.data.Facts
import java.io.File
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Proves the app's category list really comes from the facts file: the asset
 * is the verbatim data/reykjavik-form.json and must parse into all twelve
 * categories with the documented type split (almenn-abending general, the
 * other eleven specific) and the documented description limit.
 */
class FactsFileTest {

    @Test
    fun assetParsesWithAllTwelveCategories() {
        val text = File("src/main/assets/reykjavik-form.json").readText()
        val facts = Facts.parse(text)

        assertEquals(12, facts.categories.size)
        assertEquals(2500, facts.fields.description.maxLength)

        val general = facts.categories.firstOrNull { it.slug == "almenn-abending" }
        assertEquals("general", general?.type)

        val specific = facts.categories.filter { it.slug != "almenn-abending" }
        assertEquals(11, specific.size)
        assertTrue(specific.all { it.type == "specific" })

        assertTrue(facts.categories.all { it.slug.isNotBlank() && it.category.isNotBlank() && it.summary.isNotBlank() })
    }
}

File: ios/BorgarlandCore/Tests/BorgarlandCoreTests/FactsFileTest.swift
import XCTest
import BorgarlandCore

/// Proves the app's category list really comes from the facts file: the
/// package reads the verbatim data/reykjavik-form.json and it must parse into
/// all twelve categories with the documented kind split (almenn-abending
/// general, the other eleven specific) and the documented description limit.
final class FactsFileTest: XCTestCase {

    func testAssetParsesWithAllTwelveCategories() throws {
        let facts = try Facts.parse(ContractSource.dataFile("reykjavik-form.json"))

        XCTAssertEqual(facts.categories.count, 12)
        XCTAssertEqual(facts.fields.description.maxLength, 2500)

        let general = facts.categories.first { $0.slug == "almenn-abending" }
        XCTAssertEqual(general?.type, "general")

        let specific = facts.categories.filter { $0.slug != "almenn-abending" }
        XCTAssertEqual(specific.count, 11)
        XCTAssertTrue(specific.allSatisfy { $0.type == "specific" })

        XCTAssertTrue(facts.categories.allSatisfy { !$0.slug.isEmpty && !$0.category.isEmpty })
    }
}
